Add annotation to detect generated files

jsone-studios opened the pull request:
Add a `@Generated("JHipster")` annotation from `jakarta.annotation:jakarta.annotation-api` to all generated
- [x] Java classes under `src/main/java`
- [ ] Java classes under `src/test/java`

I am a bit unsure if we want to use `@Generated("JHipster")` or `@Generated("generator-jhipster")`. Any feedback would be very welcome. @pascalgrimaud 

#### Update:
A custom `GeneratedByJHipster` annotation is now added with a transform script to each java class/interface.

---

Fixes #12459

---

Please make sure the below checklist is followed for Pull Requests.

-   [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
-   [ ] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

<!--
Please also reference the issue number in a commit message to [automatically close the related GitHub issue](https://help.github.com/articles/closing-issues-via-commit-messages/)

Note: It is also possible to add `[skip ci]` or `[ci skip]` to your commit message to skip continuous integration tests
-->

pascalgrimaud commented:
Thanks @jsone-studios for this 1st draft !

About the `@Generated` annotation, I think it's better to use a custom one than the one from javax.annotation, as mentioned in the issue:

```java
package io.github.jhipster;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target({ ElementType.METHOD, ElementType.TYPE })
public @interface Generated {
}
```

Because with this, the user can customize it easily. But it's my personal opinion. It can be discussed of course.

pascalgrimaud commented:
@jsone-studios : don't make the change immediatly, let's wait some feedbacks first :)

mshima commented:
What about
```java
package io.github.jhipster;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import javax.annotation.Generated;

@Generated(string="JHipster", date="<%= creationTimestamp %>", comments="Generated by JHipster <%= jhipsterVersion %>");
@Retention(RetentionPolicy.RUNTIME)
@Target({ ElementType.TYPE })
public @interface GeneratedByJHipster {
}
```

pascalgrimaud commented:
I like it @mshima

vishal423 commented:
I think you can change the retention policy to `@Retention(RetentionPolicy.SOURCE)`. Also, instead of modifying each file to add the annotation, can we have some scripts to apply like used for prettier transformation?

mshima commented:
> I think you can change the retention policy to `@Retention(RetentionPolicy.SOURCE)`. Also, instead of modifying each file to add the annotation, can we have some scripts to apply like used for prettier transformation?

That's possible, something like:
```
 registerGeneratedByJHipsterTransform() {
        this.registerTransformStream(
            through.obj(function (file, enc, cb) {
                if (path.extname(file.path) === '.java' && file.state) {
                    file.contents = Buffer.from(file.contents.toString('utf8').replace(/\n((\w*) class)/, '@io.github.jhipster.GeneratedByJHipster\n$1', 'utf8');
                }
                this.push(file);
                cb();
            })
        );
    }
```

pascalgrimaud commented:
Don't forget the java import

jsone-studios commented:
Thanks for your intensive feedback. I really like the concept of a transformation script. I am going to refactor my PR to use such a transformation script.

gmarziou commented:
Can someone clarify what is the use of this annotation? Is this for documentation purpose (for humans) or for tools (which ones)?

pascalgrimaud commented:
@gmarziou :
- it was asked long time ago by several people, to know which files are generated by JHipster or not
- it was suggested to add a "comment" at begining of the project, but nothing has been implemented yet
- then, the idea of an annotation comes from the different projects in my company
- you can have an example here (from DDD ticket): 
    - https://gitlab.com/cdamon/padbowl/-/blob/master/src/main/java/com/jhipster/padbowl/common/infrastructure/Generated.java
    - see the commit `Mark generated classes as so (ignore coverage)` https://gitlab.com/cdamon/padbowl/-/commit/48f2d0be2752a6f7af1dd922d35917388a8edd8b
    - with this, you can completely focus on the domain and be sure to have 100% coverage on your business code

Then my first suggestion was to create a flag, something like `jhipster --generated-flag`, for people who don't want this, instead of having by default. That's why it should be in discussion.

mraible commented:
I'm not against the idea. However, since I'm not a tools vendor, how would I benefit from it? It seems like noise.

pascalgrimaud commented:
@mraible :
- as said in previous comment, it's mainly for knowing which files are generated by JHipster or not
- in real project, for real customer, after months of development, when there are new developers in team, we advise them to avoid modifying JHipster generated code
- how to know it ? they can't, unless they know JHipster, or use git log, which is annoying
- using this kind of annotation will help

jsone-studios commented:
I have pushed a new draft with a custom `GeneratedByJHipster` annotation, that is added with a transform script to every Java class and interface.

mshima commented:
I think it's possible add:
```
public @interface GeneratedByJHipster {
    /**
     * Set true to make JHipster don't override this file
     * {@code @GeneratedByJHipster(true)}
     * 
     * @return
     */
    boolean value() default false;
}
```

mshima commented on generators/app/index.js:
```diff
@@ -238,6 +238,7 @@ module.exports = class JHipsterAppGenerator extends BaseBlueprintGenerator {
 
         useBlueprints = !this.fromBlueprint && this.instantiateBlueprints('app');
 
+        this.registerGeneratedAnnotationTransform();
```
Following https://github.com/jhipster/generator-jhipster/pull/12569#issuecomment-697232896
```suggestion
        if (this.options.skipGeneratedByJhipster) {
            this.registerGeneratedAnnotationTransform();
        }
```
And add to `this.option('skip-generated-by-jhipster'...`

mshima commented on generators/generator-base-private.js:
```diff
@@ -1469,6 +1469,16 @@ module.exports = class JHipsterBasePrivateGenerator extends Generator {
     }
 
+    registerGeneratedAnnotationTransform() {
+        if (this.options.help) {
+            return;
+        }
+
```
Can be dropped.
```suggestion
```

mshima commented on generators/generator-base-private.js:
```diff
@@ -1469,6 +1469,16 @@ module.exports = class JHipsterBasePrivateGenerator extends Generator {
+            return;
+        }
+
+        const packageName = this.jhipsterConfig.packageName;
+
+        this.registerTransformStream(generatedAnnotationTransform(packageName));
```
```suggestion
        this.registerTransformStream(generatedAnnotationTransform(this.jhipsterConfig.packageName));
```

mshima commented on generators/generator-transforms.js:
```diff
@@ -50,7 +51,25 @@ const prettierTransform = function (defaultOptions) {
+const generatedAnnotationTransform = packageName => {
+    return through.obj(function (file, encoding, callback) {
+        if (path.extname(file.path) === '.java' && file.state !== 'deleted' && !file.path.endsWith('GeneratedByJHipster.java')) {
+            const content = file.contents.toString('utf-8');
```
Not sure but I think encoding should be `utf8`.

jsone-studios commented:
I have applied the suggestions from  @mshima and moved the annotation generation behind an option. I have also added support for annotating all files from the entity, sping-controller and spring-service sub-generator.

mshima commented:
> I have applied the suggestions from @mshima and moved the annotation generation behind an option. I have also added support for annotating all files from the `entity`, `sping-controller` and `spring-service` sub-generator.

You can put the option in generator-base, so every derived generator will have this option.
No need to add one by one.

In my opinion it should be opt-out instead of opt-in.

And save the option for regeneration. Like:
```
if (this.options.skip !== undefined) {
  this.jhipsterConfig.skip = ...
}
If (this.jhipsterConfig.foog.skip) {
  this.register...
}
```

mshima commented:
I think opt out could be implemented by registering `no-` option.
And applying if the option is undefined.

mshima commented on generators/app/index.js:
```diff
@@ -238,6 +238,9 @@ module.exports = class JHipsterAppGenerator extends BaseBlueprintGenerator {
 
         useBlueprints = !this.fromBlueprint && this.instantiateBlueprints('app');
 
+        if (!this.jhipsterConfig.skipGeneratedFlag) {
+            this.registerGeneratedAnnotationTransform();
+        }
```
Move to generator-base too, so it will be registered for every generator.

jsone-studios replied:
When moving this to generator-base, the `jhipster import-jdl` command does not work anymore (see failing Azure pipelines). Somehow `this.jhipsterConfig.packageName` is `undefined` at this point, but it is needed inside the `registerGeneratedAnnotationTransform()` function.

mshima replied:
We should not rely that `packageName` is defined that the constructor point.
It will never be defined when creating an app using prompts.

jsone-studios replied:
Thank you very much @mshima for your extensive help with this PR

mshima commented on generators/generator-base.js:
```diff
@@ -112,6 +112,14 @@ module.exports = class JHipsterBaseGenerator extends PrivateBase {
             this.config.set(this.options.localConfig);
         }
 
+        this.option('skip-generated-flag', {
+            desc: 'Skip adding a GeneratedByJhipster annotation to all generated java classes and interfaces',
+            type: Boolean,
+        });
```
Move the option register to before `if (this.options.help)` otherwise it won't be shown at help and commander will not be aware of this option.

mshima reviewed (changes requested):
Make the PR as ready after this changes.

mshima commented on generators/generator-base-private.js:
```diff
@@ -1469,6 +1469,13 @@ module.exports = class JHipsterBasePrivateGenerator extends Generator {
         generator.registerTransformStream([prettierFilter, prettierTransform(prettierOptions), prettierFilter.restore]);
     }
 
+    registerGeneratedAnnotationTransform(packageName) {
+        if (!packageName) {
+            packageName = this.jhipsterConfig.packageName;
+        }
+        this.registerTransformStream(generatedAnnotationTransform(packageName));
```
```suggestion
    registerGeneratedAnnotationTransform() {
        this.registerTransformStream(generatedJavaAnnotationTransform(this));
```

mshima commented on generators/generator-transforms.js:
```diff
@@ -50,7 +51,25 @@ const prettierTransform = function (defaultOptions) {
     });
 };
 
+const generatedAnnotationTransform = packageName => {
+    return through.obj(function (file, encoding, callback) {
+        if (path.extname(file.path) === '.java' && file.state !== 'deleted' && !file.path.endsWith('GeneratedByJHipster.java')) {
+            const content = file.contents.toString('utf8');
```
```suggestion
const generatedAnnotationTransform = generator => {
    return through.obj(function (file, encoding, callback) {
        if (path.extname(file.path) === '.java' && file.state !== 'deleted' && !file.path.endsWith('GeneratedByJHipster.java')) {
            const packageName = generator.jhipsterConfig.packageName;
            const content = file.contents.toString('utf8');
```

mshima commented on generators/generator-transforms.js:
```diff
@@ -50,7 +51,26 @@ const prettierTransform = function (defaultOptions) {
+        if (path.extname(file.path) === '.java' && file.state !== 'deleted' && !file.path.endsWith('GeneratedByJHipster.java')) {
+            const packageName = generator.jhipsterConfig.packageName;
+            const content = file.contents.toString('utf8');
+
```
```suggestion
if (!new RegEx(`import ${packageName.replace('\.', '\\.')}\. GeneratedByJHipster;`).test(content)) ) {
```

mshima reviewed (changes requested):
Just occurred to me that when adding needles to existing file, the annotation can be duplicated.

DamnClin commented:
Hello, 

you have to keep the retention to RUNTIME so [jacoco will ignore those files ](https://github.com/jacoco/jacoco/issues/685#issuecomment-392020612). In my opinion removing this noise is one of the main reason for this annotation (at least I'm using it for this specific reason).

jsone-studios commented:
> Hello,
> 
> you have to keep the retention to RUNTIME so [jacoco will ignore those files ](https://github.com/jacoco/jacoco/issues/685#issuecomment-392020612). In my opinion removing this noise is one of the main reason for this annotation (at least I'm using it for this specific reason).

Changing the retention to `RUNTIME` will indeed make jacoco ignore all generated files. But for a fresh generated project it might be weird for users to see an empty coverage report. Also then, no code coverage would be reported to sonar.

pascalgrimaud commented:
Then, now we have this annotation, people can easily customize with their need, so I'm fine.
Let's merge this.
Thanks a lot @jsone-studios
